prog/dftb+/lib_type/typegeometryhsd.F90

```diff
@@ -317,4 +318,129 @@ subroutine readTGeometryGen_help(node, geo, text)
+
+    !> Contains the geometry on exit
+    type(TGeometry), intent(out) :: geo
+
+    type(string) :: text
+
+    call getFirstTextChild(node, text)
+    call readTGeometryXyz_help(node, geo, char(text))
+
+  end subroutine readTGeometryXyz
+
+
+  !> Helping routine for reading geometry from a HSD tree in GEN format
+  subroutine readTGeometryXyz_help(node, geo, text)
+
+    !> Node to parse (only needed to produce proper error messages)
+    type(fnode), pointer :: node
+
+    !> Contains the geometry on exit
+    type(TGeometry), intent(out) :: geo
+
+    !> Text content of the node
+    character(len=*), intent(in) :: text
+
+    type(string) :: txt
+    integer :: iStart, iOldStart, iErr, iEnd
+    integer :: ii, iSp, iTmp
+    real(dp) :: coords(3), rTmp, det
+    type(listString) :: speciesNames
+
+    ! Read first line of the xyz file: Number of atoms
+    iStart = 1
+    iEnd = nextLine(text, iStart)
+    call getNextToken(text(:iEnd), geo%nAtom, iStart, iErr)
+    call checkError(node, iErr, "Bad number of atoms.")
+    ! xyz files are always molecular, and we are not going to support extensions to this
+    geo%tPeriodic = .false.
+    geo%tFracCoord = .false.
+
+    ! advance to next line
+    iStart = iEnd + 1
+
+    ! The parser can strip empty comment lines, so we have to try to reconstruct
+    ! the original xyz file first...
+    iOldStart = iStart
+    iEnd = nextLine(text, iStart)
+
+    ! check `second' line
+    call getNextToken(text(:iEnd), txt, iStart, iErr)
+    if (iErr == TOKEN_OK) then
+      call getNextToken(text(:iEnd), coords, iStart, iErr)
+    end if
+    if (iErr == TOKEN_OK) then
+      iStart = iOldStart  ! second line was empty or commented and therefore stripped
+    else
+      iStart = iEnd + 1  ! second line is in actual comment line, drop it
+    end if
+
+    ! Read in sequential and species indices.
+    call init(speciesNames)
+    allocate(geo%species(geo%nAtom))
+    allocate(geo%coords(3, geo%nAtom))
+    iSp = 0
+    do ii = 1, geo%nAtom
+      iEnd = nextLine(text, iStart)
+      call getNextToken(text(:iEnd), txt, iStart, iErr)
+      call checkError(node, iErr, "Bad species name for an atom.")
+      iSp = find(speciesNames, char(txt))
+      if (iSp == 0) then
+        call append(speciesNames, char(txt))
+        iSp = len(speciesNames)
+      end if
+      geo%species(ii) = iSp
+      call getNextToken(text(:iEnd), coords, iStart, iErr)
+      call checkError(node, iErr, "Bad coordinates for an atom.")
+      geo%coords(:, ii) = coords(:)
+      iStart = iEnd + 1
+    end do
+
+    geo%nSpecies = len(speciesNames)
+    allocate(geo%speciesNames(geo%nSpecies))
+    call asArray(speciesNames, geo%speciesNames)
+    call destruct(speciesNames)
+
+    if (geo%nSpecies /= maxval(geo%species) .or. minval(geo%species) /= 1) then
+      call detailedError(node, &
+          &"Nr. of species and nr. of specified elements do not match.")
+    end if
+
+    if (iStart <= len(text)) then
```

Comment:
The original  ['standard'](http://www.ccl.net/chemistry/resources/messages/1996/10/21.005-dir/index.html) does have optional scalar and vector data after the 3 coordinates, so this probably needs to be handled.
DFTB+ output xyz can contains these fields where relevant, so would need user editing for re-starts. It does however weaken the possibilities for error checking (at least each coordinate line starts with characters).
In the longer term, parsing the atomic charges and vectors could be used to simplify restarts for MD & SCC.

Reply:
In fact this is not a problem with this implementation, which comes along with the `nextLine` routine I introduced for this reader. Every token after the four columns is currently just ignored.

Reply:
I added a test example including trailing scalars in every column.